README.md

```diff
@@ -385,21 +387,46 @@ There are two built-in casting operators in **Jakt**.
+    - [ ] Float -> Integer casts truncate the decimal point, if the float is a normal numeric value. If the float is any NaN or either Infinity, the cast fails.
+    - [ ] Integer -> Float casts resolve to the closest value to the integer representable by the floating-point type. If the integer value is not exactly representable, the cast fails.
+    - [ ] Any primitive -> bool will create `true` for any value except 0, which is `false`.
+    - [ ] bool -> any primitive will do `false -> 0` and `true -> 1`, even for floats.
+- If the types are two different pointer types of the same underlying type `T`, the cast is generally allowed and infallible.
+    - [ ] A cast to `T` from any pointer type, if `T` is reference counted, will increment the reference count as expected; that's the preferred way of creating a strong reference from a weak or temporary reference.
+    - [ ] A cast to `T` from any pointer type, if `T` is *not* reference counted, will copy the value.
+    - [ ] Casts from `&T` to `&mut T` is forbidden. The other way around is allowed.
+    - [ ] A cast from and to `raw T` is unsafe but infallible.
+    - [ ] A cast to `&T` or `&mut T` is allowed for orthagonality, but has the same effect as directly taking the reference with the `&` or `&mut` operators, respectively.
+    - [ ] Casts from `&T` or `&mut T` are allowed for orthagonality and follow the same rules as dereferencing.
+    - [ ] Casts between infallible pointer types can happen alongside casts within the contained type, if the pointer type doesn't change or is cast strictly upwards on the pointer type hierarchy. This hierarchy goes (from bottom to top) `&mut T` -> `&T` -> `T`.
+- If the types are part of the same type hierarchy (i.e. one is a child type of another), casts are possible according to these rules. Note that this also goes if the types are the same pointer type and the contained type is part of the same type hierarchy (which applies recursively), so `&mut A` and `&mut B` are part of the same type hierarchy if `A` and `B` are.
+    - [ ] A child can be cast to its parent infallibly.
+    - [ ] A parent can be cast to a child, but this will check the type at runtime and fail if the object was not of the child type or one of its subtypes.
+- [ ] If the types are incompatible, a user-defined cast is attempted to be used. The details here are not decided yet, but it will involve implementing some generic trait.
+- If nothing works, the cast will not compile.
+
+Some examples of possible and not possible casts (the left side shows the data type of a `value`, not the `value` itself):
+```
+u32 as? u16  // might not succeed
+u16 as! i32  // infallible
+f32 as? u32  // normally truncates, but values like NaN cause failure
+u64 as! bool // only false if the value is 0
+A as! B      // infallible assuming A is a subclass of B
+&mut A as! A // creates a copy assuming A is a struct type; equivalent to (*value)
+i32 as! &i32 // equivalent to (&value)
```

Comment:
:yakfused: this sounds very not possible, what if the lhs is a temporary?

Reply:
In a simple implementation the C++ would probably not compile. In Rust this is allowed, you could think of an invisible `let` being introduced in the expression's statement's scope (or a smaller scope; doesn't matter). We could compile it to exactly this in C++, but it's not a priority. However, it should be allowed as I find it a useful Rust feature that stops variables from being special just because they have a name; after all you can always extract a temporary to a variable and take a reference to that. There's an argument to be had about mutability of the temporary through mutable references, but I don't think that leads to weird situations which cannot be thought of as "the temporary behaves as an implicit local variable". In most cases the mutation will just not have any effects as the pseudo-variable becomes inaccessible right after the expression evaluation.

Reply:
It's not so much about names as it is about scopes, c++ defines the scope of a value differently than that of a bound variable.

That said, I don't think we should add docs for things we haven't decided on yet?